File: TCPclient.c
/**	@file TCPclient.c
 * 	@brief Contains the function implementations of creating and running a TCP client in C. 
 *	It creates a socket for sending client connections. 
 *	It is capable of sending and receiving request to and from the TCP Server. 
 * 	@author Cole Amick
 * 	@author Daniel Davis
 * 	@bug No known bugs!
 */

#include "TCPclient.h" 


/*
 **************************************************
 *		FUNCTION PROTOTYPES
 **************************************************
 */
 
/**	@brief 	Is a function that takes in a error message and outputs the message to the display
*			and stops the server from running. 
*	@param 	Is the error message that should be outputted to the screen.
*	@return returns a -1 integer.
*/
int printErrorMessage( char *message );

/**	@brief Create a TCP socket and return the integer that represents the socket. 
*	@return the socket number. 
*/
int create_TCP_Socket(void);

/**	@brief 	get IP address of the server that the client will send messages to. 
*	@param 	serverName is the name of the server. 
*	@return returns a structure hostent that contains IP information about server. 
*/
struct hostent *info_Host(char * serverName);

/**	@brief Set the destination address and port of the server. 
*	@param 	hostptr is the structure containing information about the server 
*			port is the port number to connect to the server 
*			dest contains the destination IP address information. 
*	@return returns nothing.
*/
void setDestination(struct hostent *hostptr, int port, struct sockaddr_in *dest);

/*
 **************************************************
 *		CLIENT FUNCTIONS
 **************************************************
 */

/*
 **************************************************
 **************************************************
 */
int printErrorMessage( char *message ) {	
  fprintf(stderr, "ERROR: %s\n", message);
  return -1;
}

/*
 * Creates a streaming socket and connects to a server.
 *
 * serverName - the ip address or hostname of the server given as a string
 * port       - the port number of the server
 * dest       - the server's address information; the structure should be created with information
 *              on the server (like port, address, and family) in this function call
 *
 * return value - the socket identifier or a negative number indicating the error if a connection could not be established
 */
int createSocket(char * serverName, int port, struct sockaddr_in * dest){
	int listensockfd = 0;
	struct hostent *hostptr; 

	//create the socket
	listensockfd = create_TCP_Socket();
	if(listensockfd == -1) return -1;

	//get client host information
	hostptr = info_Host(serverName);
	if(hostptr == NULL) return -1;
	
	//set server destination
	setDestination(hostptr, port, dest);
	
	//connect to the server
	if(connect(listensockfd, (struct sockaddr *) dest, sizeof(*dest)) == -1)
		return printErrorMessage("Cannot Connect to the Server");
	
	//return the listening socket
	return listensockfd;
}


/*
 **************************************************
 **************************************************
 */
int create_TCP_Socket(void){
	int listensockfd = socket(AF_INET, SOCK_STREAM,0);
	if(listensockfd == -1)
		return printErrorMessage("Cannot Open Socket to Listen"); 
	return listensockfd;
}

/*
 **************************************************
 **************************************************
 */
struct hostent *info_Host(char * serverName){
	struct hostent *hostptr; 
	hostptr = gethostbyname(serverName);
	if(hostptr == NULL) 
		printErrorMessage("That Host Does Not Exist");			
	return hostptr;
}

/*
 **************************************************
 **************************************************
 */
void setDestination(struct hostent *hostptr, int port, struct sockaddr_in *dest){
	memset((void *) dest, 0, (size_t)sizeof(struct sockaddr_in));  
	memcpy((void *) &(dest->sin_addr), (void *) hostptr->h_addr, hostptr->h_length);
	dest->sin_family = (AF_INET);
	dest->sin_port = htons((u_short) port);
}


/*
 * Sends a request for service to the server. This is an asynchronous call to the server, 
 * so do not wait for a reply in this function.
 * 
 * sock    - the socket identifier
 * request - the request to be sent encoded as a string
 * dest    - the server's address information
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int sendRequest(int sock, char * request, struct sockaddr_in * dest){
	int error = 0;
	if(request[0] == '\0') {
		printErrorMessage("Cannot Send NULL Message to the Server");
		return -1;
	}
	error = sendto(sock, request, strlen(request), 0, (struct sockaddr *) dest, sizeof(*dest));
	if(error == -1) return -1;
	return 0;
}

/*
 * Receives the server's response formatted as an XML text string.
 *
 * sock     - the socket identifier
 * response - the server's response as an XML formatted string to be filled in by this function;
 *            memory is allocated for storing response
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int receiveResponse(int sock, char * response){
	int error = 0;
	bzero(response, MAX_MESSAGE);
	error = recvfrom(sock, response, MAX_MESSAGE, 0, NULL, NULL);
	if(error == -1) return -1;
	return 0;
}

/*
 * Prints the response to the screen in a formatted way.
 *
 * response - the server's response as an XML formatted string
 *
 */
void printResponse(char* response){
	printf("Response from server: %s\n", response); 
}

/*
 * Closes the specified socket
 *
 * sock - the ID of the socket to be closed
 * 
 * return - 0, if no error; otherwise, a negative number indicating the error
 */
int closeSocket(int sock){
	int error = close(sock);
	if(error == -1) return -1;
	return 0;
}

File: TCPclient.java
/**	@file TCPclient.java
 * 	@brief Contains the function implementations of creating and running a TCP client in JAVA.
 *	It creates a socket for sending client connections.
 *	It is capable of sending and receiving request to the TCP Server.
 * 	@author Cole Amick
 * 	@author Daniel Davis
 * 	@bug No known bugs!
 */
/*
 * TCPclient.java
 * Systems and Networks II
 * Project 1
 *
 * This file describes the functions to be implemented by the TCPclient class
 * You may also implement any auxillary functions you deem necessary.
 */
import java.net.*;
import java.io.*;
import java.util.Scanner;


public class TCPclient
{
	/*
	* Instance Fields 
	*/
	private Socket _socket; // the socket for communication with a server
    private OutputStream sendStream; //stream for sending message to the server 
    private InputStream recvStream; //stream for receiving message from the server
    
       
	/**
	 * Constructs a TCPclient object.
	 */
	public TCPclient()
	{
	}

	/**
	 * Creates a streaming socket and connects to a server.
	 *
	 * @param host - the ip or hostname of the server
	 * @param port - the port number of the server
	 *
	 * @return - 0 or a negative number describing an error code if the connection could not be established
	 */
	public int createSocket(String host, int port)
	{
        try{
        _socket = new Socket(host, port);
        sendStream = _socket.getOutputStream();
        recvStream = _socket.getInputStream();
        }
		catch(Exception ex){
        	System.out.println("Socket Stream Failed");
        	return -1;
        }
        return 0;
	}
    
	/**
	 * Sends a request for service to the server. Do not wait for a reply in this function. This will be
	 * an asynchronous call to the server.
	 *
	 * @request - the request to be sent
	 *
	 * @return - 0, if no error; otherwise, a negative number indicating the error
	 */
	public int sendRequest(String request)
	{
		if(request.length() == 0) {  //null message is attempting to be sent to the server - do not send it - return -1
			System.err.println("ERROR: Cannot Send NULL Message to the Server");
			return -1;		
        }
		try{
            byte[] sendBuff = new byte[request.length()];
            sendBuff = request.getBytes();
            sendStream.write(sendBuff, 0, sendBuff.length); //send the message to the server
        }
		catch(Exception ex){
            System.err.println("Exception in sendRequst");
        	return -1;
		}
		return 0;
	}

	/**
	 * Receives the server's response. Also displays the response in a
	 * clear and concise format.
	 *
	 * @return - the server's response or NULL if an error occured
	 */
	public String receiveResponse()
	{
		String response = "";		
		try{
            int dataSize;
            while((dataSize = recvStream.available()) == 0);
            byte[] recvBuff = new byte[dataSize];
            recvStream.read(recvBuff, 0,dataSize);
            response = new String(recvBuff,0,dataSize); //receive the message from the server
        }
		catch (Exception ex){
            System.err.println("Exception in receiveResponse");
        	return null;
		}
		return response;
	}
	
	/*
     * Prints the response to the screen in a formatted way.
     *
     * response - the server's response as an XML formatted string
     *
     */
	public static void printResponse(String response)
	{
		System.out.println("Response from server: " + response);  
	}
    
    
	/*
	 * Closes an open socket.
	 *
     * @return - 0, if no error; otherwise, a negative number indicating the error
	 */
	public int closeSocket()
	{
		try{
		    sendStream.close();  
		    recvStream.close();    	
		    _socket.close();    
		}
		catch(Exception ex){  
		    System.err.println("Exception in close"); 
			return -1;
		}
		return 0;
	}
    
	
	/**
	 * The main function. Use this function for
	 * testing your code. We will use our own main function for testing.
	 */
	public static void main(String[] args)
	{
		String response = "";
		if (args.length == 2){
		
			final String servName = args[0];
		    final int servPort = Integer.parseInt(args[1]);	
		    
			TCPclient client = new TCPclient();  
		    if ( client.createSocket(servName, servPort)  >= 0 ) //create the TCP socket 
		    {
				sendMessageTest("<echo>sfglk</echo>", response, client);
				sendMessageTest("", response, client);
				sendMessageTest("\n", response, client);
				sendMessageTest("<echo>sf\nglk</echo>", response, client);
				sendMessageTest("<echo>New Line At End</echo>\n", response, client);
				sendMessageTest("<loadavg/>", response, client);
				sendMessageTest("<echo>Hello World<echo>", response, client);
				sendMessageTest("<echo></echo>", response, client);
			
				client.closeSocket();    //close the TCP socket 
			}
		}
		else {
			System.out.println("Incorrect Number of Command Line Arguments");
			System.out.println("java TCPclient <IP Address or Server Host Name> <Port Number>");
		}
	}
	
	
	/**
	 * Used for testing the main program by sending request and receiving 
	 * responses from the server. 
	 */
	public static void sendMessageTest(String message, String response, TCPclient client) 
	{
		if( client.sendRequest(message) >= 0 ) //send a request to the server 
		{    
			response = client.receiveResponse(); //receive a response from the server
			client.printResponse(response); //print the response
		}
	}
	
}



